gap/attr.gi

```diff
@@ -290,6 +290,226 @@ function(D)
   return chrom;
 end);
 
+# Utility function to calculate the maximal independent sets (as BLists) of a
+# subgraph induced by removing a set of vertices.
+BindGlobal("DIGRAPHS_MaximalIndependentSetsSubtractedSet",
+function(I, subtracted_set, size_bound)
+  local induced_mis, temp, i;
+  # First remove all vertices in the subtracted set from each MIS
+  temp := List(I, i -> DifferenceBlist(i, subtracted_set));
+  induced_mis := [];
+  # Then remove any sets which are no longer maximal
+  # Sort in decreasing size
+  Sort(temp, {x, y} -> SizeBlist(x) > SizeBlist(y));
+  # Then check elements from back to front for if they are a subset
+  for i in temp do
+    if SizeBlist(i) <= size_bound and
+        ForAll(induced_mis, x -> not IsSubsetBlist(x, i)) then
+      Add(induced_mis, i);
+    fi;
+  od;
+  return induced_mis;
+end
+);
+
+InstallMethod(ChromaticNumber, "for a digraph and colouring algorithm",
+[IsDigraph, IsDigraphColouringAlgorithm and IsDigraphColouringAlgorithmLawler],
+function(D, Lawler)
+  local n, vertices, subset_colours, s, S, i, I, subset_iter, x,
+  MIS, subset_MIS;
+
+  n := DigraphNrVertices(D);
+  if DigraphHasLoops(D) then
+    ErrorNoReturn("the argument <D> must be a digraph with no loops,");
+  elif n = 0 then
+    return 0;  # chromatic number = 0 iff <D> has 0 verts
+  elif IsNullDigraph(D) then
+    return 1;  # chromatic number = 1 iff <D> has >= 1 verts & no edges
+  elif IsBipartiteDigraph(D) then
+    return 2;  # chromatic number = 2 iff <D> has >= 2 verts & is bipartite
+               # <D> has at least 2 vertices at this stage
+  fi;
+  vertices := List(DigraphVertices(D));
+  # Store all the Maximal Independent Sets, which can later be used for
+  # calculating the maximal independent sets of induced subgraphs.
+  MIS := DigraphMaximalIndependentSets(D);
+  # Convert each MIS to a Blist
+  MIS := List(MIS, x -> BlistList(vertices, x));
+  # Store current best colouring for each subset
+  subset_colours := ListWithIdenticalEntries(2 ^ n, infinity);
+  # Empty set can be colouring with only one colour.
+  subset_colours[1] := 0;
+  # Iterator for blist subsets.
+  subset_iter := ListWithIdenticalEntries(n, [false, true]);
+  subset_iter := IteratorOfCartesianProduct2(subset_iter);
+  # Skip the first one, which should be the empty set.
+  S := NextIterator(subset_iter);
+  # Iterate over all vertex subsets.
+  for S in subset_iter do
+    # Cartesian iterator is ascending lexicographically, but we want reverse
+    # lexicographic ordering.
+    FlipBlist(S);
```

Comment:
`FlipBlist` was added in GAP 4.11, and seems to be quite useful here.  @james-d-mitchell and @wilfwilson, would we be happy with advancing the required GAP from `>=4.9.0` to `>=4.11.0` now that it's been out for a year?  Otherwise, Ewan should provide some kind of fallback using `if IsBound(FlipBlist)`.

Reply:
Sorry I missed this last week. In short, I don't care that much, but my slight inclination is to maintain backwards compatibility with GAP 4.10 if it's not too much effort/too painful to do so, for now. However I'm happy to defer to @james-d-mitchell on this question.

Just looking briefly at the code, however, my first question is to ask whether maybe you don't need to be flipping the blist (at least not so much) - you're flipping the blist three times; is there some easy way to do your operations on the original blist directly, without having to to flip it so much?

Reply:
Having a look I've managed to eliminate a few of the flip calls, but that can't be done everywhere. For now I've just added a fallback that checks if FlipBlist can be used.

Reply:
Nice. Thanks for doing that, I hope you don't mind too much. How much slower roughly does your fallback method make your `ChromaticNumber` methods, have you done any experiments?

Reply:
I've tested it out on a few graphs, and it is worse but only marginally and it is unlikely to affect the overall performance of any code that would use this. Here is an example, just with quickly timing them (I made sure to run it a few times to check it was consistent)
```
D := DigraphRemoveLoops(RandomDigraph(15));
<immutable digraph with 15 vertices, 24 edges>
ChromaticNumber(D, DigraphColouringAlgorithmLawler);time;
3
940
MakeReadWriteGlobal("FlipBlist");
FlipBlist := function(B) Apply(B, x -> not x); end;
ChromaticNumber(D, DigraphColouringAlgorithmLawler);time;
3
962
```
For some larger graphs one went from taking 30 seconds to run to 31 seconds, so the fallback isn't too bad.